driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -405,9 +405,52 @@ private void incrementAuthErrorMetric() {
         }
     }
 
-    private UnsupportedProtocolVersionException unsupportedProtocolVersionException(ProtocolVersion triedVersion, ProtocolVersion serverProtocolVersion) {
-        logger.debug("Got unsupported protocol version error from {} for version {} server supports version {}", address, triedVersion, serverProtocolVersion);
-        return new UnsupportedProtocolVersionException(address, triedVersion, serverProtocolVersion);
+    private static final Pattern SUPPORTED_VERSIONS_PATTERN = Pattern.compile("supported versions are \\(([^)]+?)\\)");
```

Comment:
> Since C* has been pretty inconsistent about the error messages it returns

Not so much actually, I crawled through the revisions of `Frame.java` and found 4 different messages only.

> In the general case, we could simply just retry at the attempted protocol version - 1

I guess what you mean is _if_ the server version differs from the client one, then use that version, otherwise downgrade to N-1, is that right? The only problem I see with the "-1" approach is the following case: if the client sends vN and the server replies with vN + 1, but also supports vN + 2, it would be better to upgrade to vN + 2. The parsing approach would detect it, but the "-1" approach wouldn't.